[1.19]: Fix performance issue and logging when resource caching is enabled

marchermans opened the pull request:
### Fix:
- This fixes large `debug.log` files when resource caching is enabled.
- Performance degradation that occurs due to double indexing of the resource packs.

### Implemented:
- A new config loading system that properly loads the resource cache configuration data during the boot sequence
- A new logging marker to handle the large flow of potential logging data

### Changed:
- The logging infrastructure now logs all data using this marker.
- All internal classes should now be marked as being internal, making future changes easier.
- Configurations are now only registered if they are not empty.

### Removed:
- The fallback resource manager does invoke indexing when packs are pushed onto its ordered stack. The caller is responsible for enabling the pack resource initialization. _(By default vanilla does this now ahead of time anyway, just something modders need to pay attention to if they ever use the push methods on the fallback resource manager.)_

### Performance improvement
This PR fixes a major bug that relates to the fact that the index was forcefully created twice.
This was caused by the `FallBackResourceManager` invoking the initialize system twice. Once during initial registration and then once during sorting.

#### Resulting performance analysis:
After this change the numbers provided by @MarkusBordihn, which were verifyable, are now not anymore.
Using the pack he provided in #9032 I was able to verify that this fix now actually works:
**Startup time with caching disabled: 04:49**
**Startup time with caching enabled:  03:55**
Both tests were performed with several warmup runs before to filter out things like config creation etc.

### Log marker
A log marker was added to enable easy management of the logging data that is generated by the system.
By default this marker is disabled.

Closes #9032

marchermans committed:
Implement a toggle to prevent the logging of missing subsections of resource packs during caching.

marchermans committed:
Implement a toggle to prevent the logging of missing subsections of resource packs during caching.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -109,6 +110,11 @@ public static class Common {
                     .worldRestart()
                     .define("indexModPackCachesOnThread", false);
 
+            logMissingDirectoriesToDebugDuringResourceCaching = builder
+                    .comment("Set this to true to log missing directories (found during resource caching) to debug, instead of discarding the information. Note this can cause large log files in large mod packs.")
+                    .translation("forge.configgui.logMissingDirectoriesToDebugDuringResourceCaching")
+                    .worldRestart()
+                    .define("logMissingDirectoriesToDebugDuringResourceCaching", false);
```
What is the usefulness of this logging statement to debugging? As in, is this useful enough that it would merit its own dedicated config entry? 

I do not think it would be appropriate to elevate this to a Forge config setting if there is no significant usefulness for users to modify this setting. Even then, why would it not be more appropriate to use a logging marker, with the logging configuration adjusted to allow a system property to control it, [as we already do for 14 other markers](https://github.com/MinecraftForge/MinecraftForge/blob/b0caea8e204a7f026646396558d6d58a02bff9d6/fmlloader/src/main/resources/log4j2.xml#L5-L18)?

marchermans replied:
Changed it now to a marker and a different setup anyway.

marchermans committed:
Fix the reloading being stupid.
I am an idiot.
Introduce boot based forge configs which can handle boot based feature configuration.

marchermans committed:
Remove the boot configuration, it is a bit on the nose to be honest.

Fix the duplication of indexing.

marchermans committed:
Cleanup the PR

marchermans committed:
Revert the change to the CSS, I think somebody screwed up the AT.

marchermans committed:
Further clean up to maintain ABI and reduce PR size.

marchermans commented:
@sciwhiz12 Should be ready for review now.

sciwhiz12 commented on fmlloader/src/main/resources/log4j2.xml:
```diff
@@ -13,6 +13,8 @@
         <MarkerFilter marker="REGISTRIES" onMatch="${sys:forge.logging.marker.registries:-DENY}" onMismatch="NEUTRAL"/>
         <MarkerFilter marker="REGISTRYDUMP" onMatch="${sys:forge.logging.marker.registrydump:-DENY}" onMismatch="NEUTRAL"/>
         <MarkerFilter marker="SPLASH" onMatch="${sys:forge.logging.marker.splash:-DENY}" onMismatch="NEUTRAL"/>
+        <MarkerFilter marker="BOOT-CONFIG" onMatch="${sys:forge.logging.marker.boot.config:-DENY}" onMismatch="NEUTRAL"/>
```
Where is the boot config now?

sciwhiz12 commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -82,29 +82,34 @@ public static class Server {
     /**
      * General configuration that doesn't need to be synchronized but needs to be available before server startup
      */
+    @Deprecated(since = "1.19", forRemoval = true)
```
This is being deprecated in 1.19.2 (and same comment for the rest of the deprecations in here).
```suggestion
    @Deprecated(since = "1.19.2", forRemoval = true)
```

sciwhiz12 commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -82,29 +82,34 @@ public static class Server {
+        @Deprecated(since = "1.19", forRemoval = true)
         Common(ForgeConfigSpec.Builder builder) {
-            builder.comment("General configuration settings")
+            builder.comment("[DEPRECATED]:General configuration settings")
```
Changing the comments would cause the configs to be corrected I believe...? Also, the comments should reflect somewhere that these config options have no effect now, and that the configuration for resource caching has been moved to a separate config file.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -6,32 +6,32 @@
-import java.util.Map;
-import java.util.Set;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
```
No wildcard imports.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -40,8 +40,12 @@
-public class ResourceCacheManager
-{
+@ApiStatus.Internal
+public class ResourceCacheManager {
```
Please fix all of these unwanted brace changes in this class, as well as the braces in newly-added code.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,54 +73,81 @@
+        if (!Files.exists(configPath)) {
+            try {
+                Files.write(configPath, ImmutableList.of(
+                                "# This TOML configuration file controls the resource caching system which is used before the mod loading environment starts.",
+                                "# This file is read by the Forge boot loader, and is not used by the game itself.",
+                                "#",
+                                "# Set this to false to disable the resource cache. This will cause the game to scan the resource packs everytime it needs a list of resources.",
+                                "cacheResources=true",
+                                "",
+                                "# Set this to false to forge the caching of vanilla resources to happen on the main thread.",
+                                "indexVanillaPackCachesOnThread=false",
+                                "",
+                                "# Set this to false to forge the caching of mod resources to happen on the main thread.",
+                                "indexModPackCachesOnThread=false"
```
Hardcoding this to be in the manager itself and written out instead of storing a text file resource to be copied is a bit... dirty to me. [We use a resource for the FML config](https://github.com/MinecraftForge/MinecraftForge/blob/b0caea8e204a7f026646396558d6d58a02bff9d6/fmlloader/src/main/java/net/minecraftforge/fml/loading/FMLConfig.java#L40), so why not here as well?

marchermans replied:
This was what was requested by lex, not sure if it is such a good idea to hardcode.

The basic idea is to hide the fact that forge needs configs (and loads them) before mod loading as much as possible from modders, so that they do not try to load their own configs before registries are loaded.

I can make it a file, I don't care.

marchermans replied:
I will leave it a hardcode for now, should prevent people from being nut jobs and messing with it.

sciwhiz12 replied:
Could you please elaborate on what you mean by "people [...] messing with it"? Additionally, please avoid the use of language such as "nut jobs", as it can be seen as disrespectful.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,54 +73,81 @@
+
+        final List<String> lines;
+        try {
+            lines = Files.readAllLines(configPath);
```
This would mean we read the config file each time in full every time we query it. How many times do we query the config file?

marchermans replied:
Yes that is correct 🤔 
It might be worth while to cache this, but I am not 100% on that.

The invocation count is very limited and it is already faster then without caching so not sure.
If I have the time I will look into it, but that would also mean that it would only update after a reboot of the game.

marchermans replied:
I will leave it like this for now.
Invocation count in a typical gameplay loop is minimal (less then 10 if you just start the game and open a world).

This allows hot reloads in case you refresh your resource packs with F3, without requiring additional code to handle it.

If specifically requested I can build it in, but I consider that overkill for the benefit it would get.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,54 +73,81 @@
+            throw new RuntimeException("Failed to read the boot configuration file.", e);
         }
+
+        for (final String line : lines) {
+            final String trimmedLine = line.trim().replace(" ", "");
+            if (trimmedLine.startsWith(configKey)) {
+                return Boolean.parseBoolean(trimmedLine.substring(trimmedLine.indexOf('=') + 1));
+            }
+        }
```
This manual parsing of a config file pains me. Do we have no other preferred alternative?

marchermans replied:
I had the ForgeBootConfigurationManager here which just made this an ForgeConfigSpec, but lex rejected that, because it was too obvious, this was his own recommendation.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,54 +73,81 @@
+        for (final String line : lines) {
+            final String trimmedLine = line.trim().replace(" ", "");
+            if (trimmedLine.startsWith(configKey)) {
+                return Boolean.parseBoolean(trimmedLine.substring(trimmedLine.indexOf('=') + 1));
```
What would happen if the line does not contain an equals sign?

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -187,21 +214,28 @@ public Set<String> getNamespaces(final PackType type)
         return manager != null && manager.cacheLoaded();
     }
 
+    /**
+     * Functional callback interface to get a walkable stream of paths.
+     * Supports throwing {@link IOException} if the stream can not be created.
+     */
+    @FunctionalInterface
+    private interface PathWalkerFactory {
+        Stream<Path> createWalkingStream(Path path) throws IOException;
+    }
```
Why was this moved?

marchermans replied:
I originally removed the interface, and then added it back.

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -257,35 +289,31 @@ private record ResourceCacheEntry(PackType packType, String namespace, Path path
             this.packType = packType;
             this.namespace = namespace;
-            this.indexOffThread = indexOffThread;
+            this.indexOnThread = shouldIndexOnThread;
```
Why do the field and parameter not match in names?

sciwhiz12 commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -40,8 +40,12 @@
+    /**
+     * Logging marker to handle the hiding of the entries from the debug file.
+     */
+    private static final Marker RESOURCE_CACHE_MARKER = MarkerFactory.getMarker("RESOURCE-CACHE");
```
The marker does not need to be suffixed with `_MARKER`, as is done for all other markers in Forge code.
```suggestion
    private static final Marker RESOURCE_CACHE = MarkerFactory.getMarker("RESOURCE-CACHE");
```

sciwhiz12 commented on src/fmllauncher/resources/log4j2.xml:
```diff
@@ -13,6 +13,8 @@
         <MarkerFilter marker="REGISTRIES" onMatch="${sys:forge.logging.marker.registries:-DENY}" onMismatch="NEUTRAL"/>
         <MarkerFilter marker="REGISTRYDUMP" onMatch="${sys:forge.logging.marker.registrydump:-DENY}" onMismatch="NEUTRAL"/>
         <MarkerFilter marker="SPLASH" onMatch="${sys:forge.logging.marker.splash:-DENY}" onMismatch="NEUTRAL"/>
+        <MarkerFilter marker="BOOT-CONFIG" onMatch="${sys:forge.logging.marker.boot-config:-DENY}" onMismatch="NEUTRAL"/>
```
Aside from the comment about the existence (or lack thereof) of the boot config, I do note that we should really remove this `fmllauncher` folder, since I believe this is leftover from pre-PEEL.

marchermans replied:
Both config files are identical, and I am unable to determine effectively which is used at runtime, so I think it is better to keep them in sync until somebody checks them

sciwhiz12 commented on patches/minecraft/net/minecraft/server/packs/resources/FallbackResourceManager.java.patch:
```diff
@@ -1,12 +0,0 @@
- 
-    private void m_215395_(String p_215396_, @Nullable PackResources p_215397_, @Nullable Predicate<ResourceLocation> p_215398_) {
-       this.f_10599_.add(new FallbackResourceManager.PackEntry(p_215396_, p_215397_, p_215398_));
-+      if (p_215397_ != null) {
-+         p_215397_.initForNamespace(this.f_10602_);
-+      }
```
Why was this patch removed?

marchermans replied:
See PR message, it was what was causing the performance degradation.
It is not needed, since the packs are already initialized when they get to this point, causing double initialization.

marchermans committed:
Address changes requested by @sciwhiz12.

marchermans committed:
Move the PathWalkerFactory back to the bottom.

marchermans committed:
Detect when an empty configuration has been added.
Prevent adding new configurations when no options have been added.

marchermans committed:
Introduce a cache based reading of the config resource manager boot config file.
Use a file-watcher to update the cache.

Matyrobbrt commented on fmlcore/src/main/java/net/minecraftforge/fml/ModLoadingContext.java:
```diff
@@ -46,10 +51,24 @@ public <T extends Record & IExtensionPoint<T>> void registerExtensionPoint(Class
     }
 
     public void registerConfig(ModConfig.Type type, IConfigSpec<?> spec) {
+        if (spec.isEmpty())
+        {
+            // This handles the case where a mod tries to register a config, without any options configured inside it.
+            LOGGER.debug("Attempted to register an empty config for type {} on mod {}", type, getActiveContainer().getModId());
+            return;
+        }
+
         getActiveContainer().addConfig(new ModConfig(type, spec, getActiveContainer()));
     }
 
     public void registerConfig(ModConfig.Type type, IConfigSpec<?> spec, String fileName) {
+        if (spec.isEmpty())
+        {
+            // This handles the case where a mod tries to register a config, without any options configured inside it.
+            LOGGER.debug("Attempted to register an empty config for type {} on mod {} using file name {}", type, getActiveContainer().getModId(), fileName);
+            return;
+        }
+
```
These changes *can* and most probably will be considered breaking changes, so including it in an unrelated PR is not a good idea. We should try to avoid cases like #9061.

KnightMiner replied:
what mod is creating an empty config? I cannot see any practical usecase.

marchermans replied:
This was requested by lex.

Matyrobbrt commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -83,28 +83,31 @@ public static class Server {
 
         Common(ForgeConfigSpec.Builder builder) {
-            builder.comment("General configuration settings")
+            builder.comment("[DEPRECATED / NO EFFECT]:General configuration settings")
```
```suggestion
            builder.comment("[DEPRECATED / NO EFFECT]: General configuration settings")
```
Spacing for readability.

Matyrobbrt commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -83,28 +83,31 @@ public static class Server {
 
             cachePackAccess = builder
-                    .comment("Set this to true to cache resource listings in resource and data packs")
+                    .comment("[DEPRECATED / NO EFFECT] [NOW IN BOOT CONFIG]:Set this to true to cache resource listings in resource and data packs")
```
```suggestion
                    .comment("[DEPRECATED / NO EFFECT] [NOW IN BOOT CONFIG]: Set this to true to cache resource listings in resource and data packs")
```
Spacing for readability.

Matyrobbrt commented on src/main/java/net/minecraftforge/common/ForgeConfig.java:
```diff
@@ -83,28 +83,31 @@ public static class Server {
 
             indexVanillaPackCachesOnThread = builder
-                    .comment("Set this to true to index vanilla resource and data packs on thread")
+                    .comment("[DEPRECATED / NO EFFECT] [NOW IN BOOT CONFIG]:Set this to true to index vanilla resource and data packs on thread")
```
```suggestion
                    .comment("[DEPRECATED / NO EFFECT] [NOW IN BOOT CONFIG]: Set this to true to index vanilla resource and data packs on thread")
```
Spacing for readability.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -51,7 +71,7 @@
     /**
      * Indicates if the indexing of the file-tree should happen off-thread or on-thread and block the process accordingly.
```
Javadoc no longer matches field usage.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,35 +89,39 @@
+        this.pathBuilder = pathBuilder;
     }
 
+    /**
+     * Indicates if the caching system is enabled or not.
+     *
+     * @return {@code true} if the caching system is enabled, {@code false} otherwise.
+     */
```
```suggestion
    /**
     * {@return if the caching system is enabled}
     */
```

marchermans replied:
There is a significant difference between the returns and the method body documentation in that it describes the returned values.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -235,7 +259,7 @@ private record ResourceCacheEntry(PackType packType, String namespace, Path path
         /**
          * Indicates if the indexing should run off-thread.
```
Javadoc no longer matches field usage.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -413,6 +446,178 @@ public boolean cacheLoaded()
+         * Sets up, and runs a monitor on the boot configuration file, on the current thread.
+         * Make sure to invoke this from a background thread.
+         */
+        private static void monitorBootConfiguration()
+        {
+            //Get the configuration directory.
+            Path path = CONFIG_PATH.getParent();
+            try (final WatchService watchService = FileSystems.getDefault().newWatchService())
+            {
+                //Register a watch service on the configuration directory for any changes.
+                path.register(watchService, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
+                while (true)
+                {
+                    //Grab the next key
+                    WatchKey key = watchService.take();
+                    //For each event in that key we check if it is our config file that is changed.
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        if (event.context().equals(path.relativize(CONFIG_PATH)))
+                        {
+                            //And then we atomically update the config contents.
+                            CURRENT_CONFIG_CONTENTS.set(readConfigFileContent());
+                        }
+                    }
+                    //Reset the key, and continue.
+                    key.reset();
+                }
+            }
+            catch (IOException | InterruptedException e)
+            {
+                //Some generic failure occurred, that is not good. Let's kill it!
+                throw new RuntimeException("Failed to read boot configuration contents on update.", e);
+            }
+        }
```
This can be changed to use the nightconfig-provided `FileWatcher`, reinventing the wheel is not needed.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -413,6 +446,178 @@ public boolean cacheLoaded()
+            {
+                //For each line: Remove all strings, and then check if it is the line with the requested key value.
+                final String trimmedLine = line.trim().replace(" ", "");
+                if (trimmedLine.startsWith("%s=".formatted(configKey))) //We look for key=value combindation in the trimmed
```
~~According to the [TOML Spec](https://toml.io/en/v1.0.0#keyvalue-pair), which as it stands, this bare-bones reader does not follow:
Whitespace is ignored around key names and values.
Additionally, if as you said, the reading of the file happens few times, I do not see a problem with using nightconfig to read it.~~
Disregard, didn't see the `replace` call.

marchermans replied:
Misread by @Matyrobbrt Tested the base file in toml lint.
Works.

Matyrobbrt commented on src/main/java/net/minecraftforge/resource/ResourceCacheManager.java:
```diff
@@ -69,35 +89,39 @@
         this.pathBuilder = pathBuilder;
     }
 
-    public boolean shouldIndexOffThread()
```
Removal of this method is a binary breaking change, as previously the class had not been considered internal. It should get re-added, `@Deprecated(forRemoval)` and it would return the inversed value of `shouldIndexOnThread`

eerussianguy replied:
fwiw this was not considered public api when it was reverted in 1.18. So the core team may be taking the same view here.

KnightMiner replied:
That whole revert was a breaking change. This PR sounds like its fixing a bug, not making a revert.

marchermans committed:
Address requested changes by Maty

marchermans committed:
Use NightConfig

marchermans committed:
Fix comment

marchermans committed:
Flip the condition.
